### PR TITLE
Add retag method

### PR DESCRIPTION
**Pull request by matter123**
This PR adds the reTag feature as discussed in #181

`reTag` allows duplicating a pattern while mapping a tag name, reference, or group number to a new tag name. If the special keys "all", or "keep" are set to true, unchanged tags are kept, otherwise, they are removed. This does preclude using "all" or "keep" as key for re-tagging purposes.

`reTag` does not modify the original pattern and preserves group numbers and references.

Example Usage:
```ruby
grammar[:original] = original = newPattern(
    match: /new/.then(
        match: /\s?pattern/,
        tag_as: "entity.name.inner",
        reference: "inner"
    ).then(
        match: /\s?test/,
        reference: "bar"
    ),
    tag_as: "entity.name.outer"
)

grammar[:new] = original.reTag(
    "entity.name.inner" => "entity.name.retagged",
    "all" => true
)
```
The above example produces the following grammar fragment:
```json
"original": {
    "match": "new(\\s?pattern)(\\s?test)",
    "captures": {
    "1": {
        "name": "entity.name.inner.retag"
    }
    },
    "name": "entity.name.outer.retag"
},
"new": {
    "match": "new(\\s?pattern)(\\s?test)",
    "captures": {
    "1": {
        "name": "entity.name.retagged.retag"
    }
    },
    "name": "entity.name.outer.retag"
}
```

Resolves: #181
Part of: #290

### CODE REVIEW

**Comment by jeff-hykin** [comment]
to differentiate the retag name from the options I changed all and keep to symbols: 
```
grammar[:new] = original.reTag(
    "entity.name.inner" => "entity.name.retagged",
    all: true,
)
```

**Comment by jeff-hykin** [comment]
I kind of think that the default value for all/keep should be true, and then it can manually be set to false. I think most of the time the retagging will just be to add additional specificity to some (or all) of the tags.  I doubt there will be many times when the tags will need to be removed

**Comment by matter123** [comment]
Yeah, I was just thinking that. Should be easy enough to default keep_tags to true. Somethink like `keep_tags = not arguments[:all] == false or arguments[:keep] == false`

File: source/textmate_tools.rb [hunk 1]
```diff
@@ -539,7 +539,38 @@ def backReference(reference)
         new_regex.group_attributes = self.group_attributes
         return new_regex
     end
-    
+    def reTag(arguments)
+        keep_tags = arguments[:all] == true or arguments[:keep] == true
+        self_as_string = self.without_default_mode_modifiers
+        new_regex = /#{self_as_string}/
+        new_attributes = Marshal.load(Marshal.dump(self.group_attributes))
+        # this is O(N*M) and could be expensive if reTagging a bit pattern
```
Comment (matter123) [inline source/textmate_tools.rb]:
```suggestion
        # this is O(N*M) and could be expensive if reTagging a big pattern
```

File: source/textmate_tools.rb [hunk 2]
```diff
@@ -539,7 +539,38 @@ def backReference(reference)
         new_regex.group_attributes = self.group_attributes
         return new_regex
     end
-    
+    def reTag(arguments)
+        keep_tags = arguments[:all] == true or arguments[:keep] == true
```
Comment (matter123) [inline source/textmate_tools.rb]:
```suggestion
        keep_tags = not (arguments[:all] == false or arguments[:keep] == false)
```

**Comment by jeff-hykin** [comment]
If you want to, I think a great tweak to this would be an `:append` option. reTag(append: ".template") where every tag inside the pattern just gets additional specificity. I'd implement it but I'm not exactly sure how the reTag function works.

**Comment by matter123** [comment]
That sounds like a great tweak. I'll add it. Does append apply to explicitly retagged scopes?

That is given `retag("foo" => "bar", append: "baz")` does foo become bar or bar.baz?

**Comment by matter123** [comment]
The basic method for how reTag works is:
1. make a deep clone of the group attributes
2. for every group attribute:
   1. if there is no tag_as assume the group is for the refernce and preserve it
   2. for each argument:
      1. compare the key to the tag, reference and group number, if there is a match set tag_as to the value and mark the group for preserving
3. if keep_tags is false remove tag_as from the keys not marked as preserved
4. remove the preserved markings
5. construct and return the new regex

**Comment by jeff-hykin** [comment]
> That sounds like a great tweak. I'll add it. Does append apply to explicitly retagged scopes?

Just saw this, the way you implemented it was the way I was thinking

**Comment by jeff-hykin** [comment]
I tweaked it a bit to use a `__deep_clone__` method and I added retagging support to PatternRanges

**Pull request merged**